### PR TITLE
Fixes the fair sampling, but not entirely sure; 

### PR DESCRIPTION
**Pull request by yjojo17**
Makes the Missing Value Parameter dependet on 50% of largest column.
And fixes the fair sampling but not sure if the logic is fault tolerant now, how I made it. Is not that easy to undestand what i do there so if you want to go over the PR just contact me.

### CODE REVIEW

**Comment by ClFeSc** [comment]
> Is not that easy to undestand what i do there so if you want to go over the PR just contact me.

I agree with that. I haven't really understood it, so I'll look into it again later.

File: pysrc/models/metanome_run.py [hunk 1]
```diff
@@ -293,15 +294,16 @@ def parse_results(result_file_name: str, *, algorithm: str, arity: str, results_
...
+def run_metanome(configuration: MetanomeRunConfiguration, output_fname: str, pipe: bool, largest_unique_count: int) -> MetanomeRun:
     # TODO: Make these configurable
     if configuration.algorithm == 'BINDER':
         algorithm_path = 'BINDER.jar'
         algorithm_class_name = 'de.metanome.algorithms.binder.BINDERFile'
     elif configuration.algorithm == 'PartialSPIDER':
         algorithm_path = 'PartialSPIDER.jar'
         algorithm_class_name = 'de.metanome.algorithms.spider.SPIDERFile'
-        missing_values = 10000 # TODO: Make this configurable and dependent on the sample size
+        #TODO Probably need the value of the parameter also in the JSON output or?
```
Comment (ClFeSc) [inline pysrc/models/metanome_run.py]: I have refactored that.

File: pysrc/scripts/sampling.py [hunk 2]
```diff
@@ -41,17 +41,34 @@ def assign_budget(size_per_column: list[list[ColumnBudgetInfo]], budget_to_share
         if not size_for_column.full_column_fits_in_budget
     )
 
-    if count_columns_not_full == 0:
+    if count_columns_not_full <= 1:
```
Comment (ClFeSc) [inline pysrc/scripts/sampling.py]: Just curious: Was this an actual problem?

File: pysrc/scripts/sampling.py [hunk 3]
```diff
@@ -245,19 +260,20 @@ def run_experiments(dataset: str, config: GlobalConfiguration) -> str:
     for sampling_method in config.sampling_methods:
         for budget in config.total_budget:
             new_file_list: list[tuple[str, str, int]] = []
+            track_changes = 0
```
Comment (ClFeSc) [inline pysrc/scripts/sampling.py]: Any particular reason for declaring this here, and not just passing a literal `0` to `assign_budget`?

**Review by ClFeSc — COMMENTED** [review]
I've taken a deeper look, refactored the `missing_values` count a bit, and merged the current changes. Still have to understand the `assign_budget` method.

File: pysrc/scripts/sampling.py (hunk 3 above)
Comment (yjojo17) [reply pysrc/scripts/sampling.py]: nope don't think I have a reason for it will adapt this

File: pysrc/scripts/sampling.py [hunk 4]
```diff
@@ -40,33 +41,42 @@ def assign_budget(size_per_column: list[list[ColumnBudgetInfo]], budget_to_share
...
 
-
+    # budget that was left after the initial phase, which is now distributed between the columns that require more budget
     budget_per_column = math.floor(budget_to_share / count_columns_not_full)
 
     for sizes_for_file in size_per_column:
         for size_for_column in sizes_for_file:
             if not size_for_column.full_column_fits_in_budget:
+                # check if the amount of unique values fits now with the additional budget then the allowed budget parameter
+                # it doesn't fit so far but the budget of this iteration is added to the basic size for the next iteration
```
Comment (ClFeSc) [inline pysrc/scripts/sampling.py:65]: I don't get this comment. `basic_size` is not even assigned in the next few lines which I'd have expected from `the budget of this iteration is added to the basic size for the next iteration`.

File: pysrc/scripts/sampling.py [hunk 5]
```diff
@@ -40,33 +41,42 @@ def assign_budget(size_per_column: list[list[ColumnBudgetInfo]], budget_to_share
...
                 if size_for_column.allowed_budget > budget_per_column + basic_size:
                     #size_for_column.allowed_budget = budget_per_column + basic_size
                     budget_to_share -= budget_per_column
+
+                # decrease the budget pool by the now used budget
```
Comment (ClFeSc) [inline pysrc/scripts/sampling.py]: You say `decrease` here but `budget_to_share` will be larger after the `else`.

**Review by ClFeSc — COMMENTED** [review]
I'm still not sure I understand everything, but it's much better now.

File: pysrc/scripts/sampling.py [hunk 6]
```diff
@@ -40,17 +41,45 @@ def assign_budget(size_per_column: list[list[ColumnBudgetInfo]], budget_to_share
...
     for sizes_for_file in size_per_column:
         for size_for_column in sizes_for_file:
             if not size_for_column.full_column_fits_in_budget:
-                size_for_column.allowed_budget += budget_per_column
+                # check if the amount of unique values fits now with the additional budget then the allowed budget parameter
+                # it doesn't fit so far but the budget of this iteration is added to the basic size for the next iteration
+                # the adding is performed in line 82
+                # decrease the budget pool by the now used budget
+                if size_for_column.allowed_budget > budget_per_column + basic_size:
+                    #size_for_column.allowed_budget = budget_per_column + basic_size
```
Comment (ClFeSc) [inline pysrc/scripts/sampling.py:69]: Is this line still relevant?

**Review by ClFeSc — COMMENTED** [review]
I think I'm starting to understand what you're doing. @dkuska What do you think about this?

**Comment by ClFeSc** [comment]
Was https://github.com/dkuska/hpi-adp-ind/pull/50/commits/20b59b44c40bee069a732f88d916f5c9260b538b intended by any means? It seems ... destructive.

**Comment by yjojo17** [comment]
> 

No my Version Control got messed up my VM died while pushing trying to resolve it currently

**Comment by yjojo17** [comment]
In FixFairSamplin2 is the corrected one

**Comment by yjojo17** [comment]
> 

Interestingly the commit before the destructive one had already all relevant changes not sure why git wasn't me allowing to only commit this ._.

**Comment by yjojo17** [comment]
Approve?

**Review by ClFeSc — CHANGES_REQUESTED** [review]
There are still at least four problems:

1. The budget is not used as efficiently as possible: In SCOP, there exists a distribution that uses 478 of the budget (7 and 11 for the columns that fit entirely and 23 for the rest). In case the total budget is at least 478 but not more than 483, this distribution is not chosen. Instead, 22 for all columns that don't fit is chosen. This is however not a large range where it is incorrect, and it stays under budget.
2. In SCOP, when the allowed budget is under 418, the budget is not respected and every column seems to be entirely taken. This is also not that bad as this only happens for small budgets.
3. When choosing a budget that is too large (e.g. 1 million for SCOP), I get a division by zero error.
4. There are still budgets where the distribution exceeds the budget, e.g. 500,000 for SCOP.

**Comment by yjojo17** [comment]
> 

1. is probably caused thgrough math.floor in calculations of the budgets 
2. hit a break condition that shouldn't be hit is fixed now
3. is fixed missed to a push a local commit therere
4.  is also fixed i returned to much budget to the pool in the recursive calling of the function.

**Review by ClFeSc — APPROVED** [review]
New changes seem to be working.

There is still the issue that in some cases there might be a more efficient way of assigning budget, but what I could find were only very small deviations, so from my point of view it's fine.

Also, there are some code-style imperfections in the latest commit, but I'd be fine with keeping them for the sake of getting this ready.

**Pull request merged**